services/paymaster/src/api.rs

```diff
@@ -157,6 +172,231 @@ pub async fn validate_transaction(
     Ok(())
 }
 
+pub fn validate_transaction_against_variation(
+    transaction: &VersionedTransaction,
+    tx_variation: &TransactionVariation,
+    sponsor: &Pubkey,
+) -> Result<(), (StatusCode, String)> {
+    match tx_variation {
+        TransactionVariation::V1(variation) => validate_transaction_against_variation_v1(transaction, variation, sponsor),
+    }
+}
+
+// TODO: incorporate gas spend and rate limit checks
+pub fn validate_transaction_against_variation_v1(
+    transaction: &VersionedTransaction,
+    variation: &crate::constraint::TransactionVariationV1,
+    sponsor: &Pubkey,
+) -> Result<(), (StatusCode, String)> {
+    let instructions = transaction.message.instructions();
+    if instructions.len() != variation.instructions.len() {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            format!(
+                "Transaction instruction count {} does not match expected count {} for variation {}",
+                instructions.len(),
+                variation.instructions.len(),
+                variation.name
+            ),
+        ));
+    }
+
+    for (instr, constraint) in instructions.iter().zip(&variation.instructions) {
+        let program_id = instr.program_id(transaction.message.static_account_keys());
+        if *program_id != constraint.program {
+            return Err((
+                StatusCode::BAD_REQUEST,
+                format!(
+                    "Transaction instruction program ID {} does not match expected ID {} for variation {}",
+                    program_id,
+                    constraint.program,
+                    variation.name
+                ),
+            ));
+        }
```

Comment:
Is there any reason we might want to collect all the errors rather than just returning the first we encounter?